User email preferences service migration from angularjs to angular fixes part of (#8472)

ashutoshc8101 opened the pull request:
# Overview
1. This PR fixes part of #8472
2. This PR does the following: Migrate user-email-preferences.service.ts to angular

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
https://user-images.githubusercontent.com/24855641/108723829-af0fa580-754a-11eb-8412-792dbb4eda55.mp4


![Screenshot from 2021-02-22 20-13-56](https://user-images.githubusercontent.com/24855641/108723643-838cbb00-754a-11eb-9cce-dc44c671d592.png)


## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

ashutoshc8101 committed:
Migration complete

ashutoshc8101 committed:
Two left

ashutoshc8101 committed:
Complete

ashutoshc8101 committed:
Changed to backend api service

ashutoshc8101 committed:
Antoher one

ashutoshc8101 committed:
another one

ashutoshc8101 committed:
sd

ashutoshc8101 committed:
sfd

ashutoshc8101 committed:
sf

ashutoshc8101 committed:
Final

ashutoshc8101 committed:
Final 2

ashutoshc8101 committed:
sdf

ashutoshc8101 commented:
@kevintab95 @nithusha21 @srijanreddy98 PTAL

srijanreddy98 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences-backend-api.service.ts:
```diff
@@ -0,0 +1,103 @@
+      suggestionNotificationsMuted: boolean): void {
+    this.feedbackNotificationsMuted = feedbackNotificationsMuted;
+    this.suggestionNotificationsMuted = suggestionNotificationsMuted;
+  }
+  /**
+   * @return {boolean} Whether the feedback notification is muted.
+   */
+  areFeedbackNotificationsMuted(): boolean {
+    return this.feedbackNotificationsMuted;
+  }
+  /**
+   * @return {boolean} Whether the suggestion notification is muted.
+   */
+  areSuggestionNotificationsMuted(): boolean {
+    return this.suggestionNotificationsMuted;
+  }
+  /**
+   * Set the message type to feedback and mute to true or false.
+   * @param {boolean} mute - Whether the feedback notification is muted.
+   */
+  setFeedbackNotificationPreferences(mute: boolean): void {
+    this.saveChangeToBackend({
+      message_type: this.MESSAGE_TYPE_FEEDBACK,
+      mute: mute
+    });
+  }
+  /**
+   * Set the message type to suggestion and mute to true or false.
+   * @param {boolean} mute - Whether the suggestion notification is muted.
+   */
+  setSuggestionNotificationPreferences(mute: boolean): void {
+    this.saveChangeToBackend({
+      message_type: this.MESSAGE_TYPE_SUGGESTION,
+      mute: mute
+    });
+  }
+  /**
+   * Save the change of message_type and mute to backend.
+   * @param {object} requestParams - Info about message_type and mute.
+   */
```
These should be in a file called user-email-preference-backend-api.service.ts. Here is a quick tip to differentiate between what should go in backend API service: Only those functions that directly use httpClient usually end up in backendapi service.

ashutoshc8101 replied:
Done

srijanreddy98 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences-backend-api.service.ts:
```diff
@@ -0,0 +1,103 @@
+  saveChangeToBackend(
+      requestParams: object
+  ): Promise<void | object> {
+    var that = this;
```
Why do you need to do `var that = this`? If you use `() => {}` instead of `function () {}` in Line 91 you don't need to do what has been done here. Also, use let/ const instead of var here and everywhere.

ashutoshc8101 replied:
It was done like this before migration. I will fix this.

ashutoshc8101 replied:
Fixed

ashutoshc8101 replied:
Fixed

srijanreddy98 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences-backend-api.service.ts:
```diff
@@ -0,0 +1,103 @@
+ * @fileoverview User exploration emails service for the exploration settings.
+ */
+
+interface ResponseData {
```
The name of the interface is good. But can we make it better? Maybe something more explicit to what kind of data it is representing.

ashutoshc8101 replied:
EmailPreferencesData is good??

ashutoshc8101 replied:
EmailPreferencesData is good??

srijanreddy98 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences-backend-api.service.ts:
```diff
@@ -0,0 +1,103 @@
+   * @param {object} requestParams - Info about message_type and mute.
+   */
+  saveChangeToBackend(
+      requestParams: object
```
The type of Object doesn't really specify the structure of the requestParams. Can you try and find out what could be the structure of the payload here? To get the type, you could go to the backend route handler and infer the type from there.

ashutoshc8101 replied:
Fixed.

ashutoshc8101 replied:
Fixed.

srijanreddy98 reviewed (changes requested):
Thanks, @ashutoshc8101! Good work but needs some improvements. Took a pass.

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into UserEmailPreferencesServiceMigration

ashutoshc8101 committed:
Wrong implementation reverted

ashutoshc8101 committed:
Refactored

ashutoshc8101 committed:
File overview added

ashutoshc8101 committed:
undo changes to settings-tab

ashutoshc8101 committed:
undo settings-tab 2

ashutoshc8101 committed:
undo exploration-editor-page

ashutoshc8101 committed:
undo setting tab 3

ashutoshc8101 commented:
@srijanreddy98  PTAL

nithusha21 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences-backend-api.spec.ts:
```diff
@@ -0,0 +1,66 @@
+/**
+ * @fileoverview Unit tests for the UserEmailPreferencesBackendApiService.
+ */
+import { UserEmailPreferencesBackendApiService } from './user-email-preferences-backend-api.service';
```
Add newlines above

ashutoshc8101 replied:
Fixed.

nithusha21 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences-backend-api.spec.ts:
```diff
@@ -0,0 +1,66 @@
+ */
+import { UserEmailPreferencesBackendApiService } from './user-email-preferences-backend-api.service';
+import { HttpClientTestingModule, HttpTestingController } from
+  '@angular/common/http/testing';
```
You can put this on the same line as the import statement

nithusha21 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences-backend-api.spec.ts:
```diff
@@ -0,0 +1,66 @@
+    }
+  };
+
+  var serviceInstance: UserEmailPreferencesBackendApiService = null;
```
Don't need to initialize to null, and probably change var to let.

ashutoshc8101 replied:
Fixed

nithusha21 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences.service.spec.ts:
```diff
@@ -36,81 +34,79 @@ describe('User Email Preferences Service', function() {
-    angular.mock.module('oppia');
+  var serviceInstance: UserEmailPreferencesService = null;
+  var httpTestingController: HttpTestingController = null;
+  var csrfService: CsrfTokenService = null;
```
I think generally, the naming is done where the variable name is the same as the service, with just the first letter as small.  Seems to have been incorrect in the original code itself, can we fix that?

ashutoshc8101 replied:
Can you please, elaborate more? Do you want to change serviceInstance to userEmailPreferencesService or make instance's name and service's name same?

ashutoshc8101 replied:
Fixed.

nithusha21 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences.service.spec.ts:
```diff
@@ -36,81 +34,79 @@ describe('User Email Preferences Service', function() {
 
-  it('should successfully intialise the service', function() {
-    expect(UserEmailPreferencesService.feedbackNotificationsMuted)
+
```
Extra newline?

ashutoshc8101 replied:
Fixed.

nithusha21 reviewed (commented):
Took a pass over the spec files!

ashutoshc8101 committed:
Requested changes are resolved

ashutoshc8101 commented:
@srijanreddy98 @nithusha21 @kevintab95 PTAL

kevintab95 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences-backend-api.service.ts:
```diff
@@ -0,0 +1,44 @@
+  saveChangeToBackend(
+      requestParams: RequestParams
+  ): Promise<void | object> {
+    var emailPreferencesUrl = '/createhandler/notificationpreferences/12345';
```
This seems incorrect. What does 12345 represent?

ashutoshc8101 replied:
Yes, that was a mistake. Fixed it now.

kevintab95 reviewed (changes requested):
Thanks @ashutoshc8101! Left a comment. PTAL!

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into UserEmailPreferencesServiceMigration

ashutoshc8101 committed:
Tests are fixed

ashutoshc8101 commented:
@kevintab95 @nithusha21 @srijanreddy98 PTAL

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into UserEmailPreferencesServiceMigration

kevintab95 reviewed (changes requested):
Thanks @ashutoshc8101! Can you please confirm that you have tested this functionality manually? If so, please update the "Proof that changes are correct" section in the PR description accordingly.

ashutoshc8101 commented:
@kevintab95 PTAL.
Thank You for your review.
I have added appropriate screenshot and video as proof of correct changes as suggested.

kevintab95 commented:
Hi @ashutoshc8101, in the video, it seems like you had to click outside the button after toggling it for the value to update. Is it correct?

ashutoshc8101 commented:
@kevintab95 Yes, I have to click outside the button to update the text.

kevintab95 commented:
Ah, I think that is unexpected. Looks like before your changes, it worked without a click outside.

Perhaps you need to call `$rootscope.$applyAsync()` in the successCallback of async methods? (see 4th question in [FAQ](https://github.com/oppia/oppia/wiki/Angular-Migration#faq))

ashutoshc8101 committed:
Fixed

ashutoshc8101 commented:
@kevintab95 Thank you for your suggestion.
But alone using `$rootScope.$applyAsync()` does not fix this issue.
So, I used a timeout, after much hit and trial, I found 150ms to be the minimum time that can be used to solve this issue.
PTAL.

ashutoshc8101 commented:
Proof.
Uploading threadTable-2021-02-22_18.57.02.mp4…

ashutoshc8101 commented:
@kevintab95 PTAL.
Proof:

![threadTable-2021-02-22_18 57 02](https://user-images.githubusercontent.com/24855641/108715056-59360000-7540-11eb-9539-38928a739c2e.gif)

kevintab95 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences.service.ts:
```diff
@@ -12,81 +12,94 @@
+   * Save the change of message_type and mute to backend.
+   * @param {RequestParam} requestParams - Info about message_type and mute.
+   */
+  saveChangeToBackend(requestParams: RequestParams) : Promise<void | object> {
```
Hi @ashutoshc8101, I think what needs to be done is, you need to pass in a successCallback here and call that once `saveChangeToBackend` has resolved (i.e. at the end of all the statements inside the `then` block).

This successCallback will be a function that executes `$rootscope.$applyAsync()`. You will need to pass this function in all the places where `UserEmailPreferencesService.saveChangeToBackend` is called.

ashutoshc8101 replied:
Done.

kevintab95 reviewed (changes requested):
Hi @ashutoshc8101 please avoid using timeouts unless it is absolutely required because this can cause our tests to be come a bit flaky. I've left more details in an inline comment.

ashutoshc8101 committed:
callback

ashutoshc8101 commented:
@kevintab95 PTAL.
Thank you for your review and suggestion.
I have added successCallback.
Proof:

https://user-images.githubusercontent.com/24855641/108723829-af0fa580-754a-11eb-8412-792dbb4eda55.mp4


![Screenshot from 2021-02-22 20-13-56](https://user-images.githubusercontent.com/24855641/108723643-838cbb00-754a-11eb-9cce-dc44c671d592.png)

ashutoshc8101 commented:
@kevintab95 @srijanreddy98 Please review.

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into UserEmailPreferencesServiceMigration

ashutoshc8101 committed:
fixed setting-tab.component.spec.ts

ashutoshc8101 commented:
I have updated `settings-tab.component.spec.ts` because in previous PR (merged) (#11930) , I have added one more parameter to `setFeedbackNotificationPreferences` and `setSuggestionNotificationPreferences`.

srijanreddy98 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences-backend-api.service.ts:
```diff
@@ -0,0 +1,52 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { UrlInterpolationService } from 'domain/utilities/url-interpolation.service';
+import { ExplorationDataService } from './exploration-data.service';
+import { RequestParams } from './user-email-preferences.service';
```
Move these imports below the file overview, here and every other file.

ashutoshc8101 replied:
Done.

srijanreddy98 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences-backend-api.service.ts:
```diff
@@ -0,0 +1,52 @@
+
+  saveChangeToBackend(
+      requestParams: RequestParams
+  ): Promise<void | object> {
```
Can you find out if this returns void or returns an Object. If it returns an Object, find the exact type. Object is too generic.

ashutoshc8101 replied:
Done.

ashutoshc8101 replied:
I found the exact type.
But lint tests fails when I use a type other an object.
Everything works as expected on the web app.
What should I do disable lint checks for this or use the type object.
![Screenshot from 2021-02-25 12-59-05](https://user-images.githubusercontent.com/24855641/109118481-61bd4f00-7769-11eb-9f0a-d700e79c7ec0.png)

ashutoshc8101 replied:
Done.

srijanreddy98 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences-backend-api.spec.ts:
```diff
@@ -0,0 +1,75 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for the UserEmailPreferencesBackendApiService.
```
Rename this file from `user-email-preferences-backend-api.spec.ts` to `user-email-preferences-backend-api.service.spec.ts`

ashutoshc8101 replied:
Done

srijanreddy98 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences.service.spec.ts:
```diff
@@ -18,99 +18,103 @@
+    fakeAsync(() => {
+      userEmailPreferencesService
+        .setFeedbackNotificationPreferences(false, () => {});
+      var req = httpTestingController.expectOne(
```
Use `let/ const` everywhere

ashutoshc8101 replied:
Done.

srijanreddy98 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences.service.ts:
```diff
@@ -12,81 +12,100 @@
+   * Save the change of message_type and mute to backend.
+   * @param {RequestParam} requestParams - Info about message_type and mute.
+   */
+  saveChangeToBackend(requestParams: RequestParams) : Promise<void | object> {
```
Promise<void | object> is not specific enough for what might be returned. Can you find the exact type?

ashutoshc8101 replied:
Done

ashutoshc8101 replied:
I found the exact type.
But lint tests fails when I use a type other an object.
Everything works as expected on the web app.
What should I do dI found the exact type.
But lint tests fails when I use a type other an object.
Everything works as expected on the web app.
What should I do disable lint checks for this or use the type object.
![Screenshot from 2021-02-25 12-59-05](https://user-images.githubusercontent.com/24855641/109118481-61bd4f00-7769-11eb-9f0a-d700e79c7ec0.png)
isable lint checks for this or use the type object.
![Screenshot from 2021-02-25 12-59-05](https://user-images.githubusercontent.com/24855641/109118481-61bd4f00-7769-11eb-9f0a-d700e79c7ec0.png)

ashutoshc8101 replied:
Done.

srijanreddy98 reviewed (changes requested):
Thanks @ashutoshc8101. Took a pass. PTAL!

ashutoshc8101 committed:
Suggested changes

ashutoshc8101 commented:
@srijanreddy98 PTAL.
I want implemented the suggested changes.
Everything works as expected.

https://user-images.githubusercontent.com/24855641/109119562-dc3a9e80-776a-11eb-9caa-d39e6f0e2e0c.mp4

But lint tests are failing.
![Screenshot from 2021-02-25 12-59-05](https://user-images.githubusercontent.com/24855641/109119608-ebb9e780-776a-11eb-89ce-a4c894dbdb06.png)

darksun27 commented on core/templates/services/angular-services.index.ts:
```diff
@@ -337,6 +337,9 @@ import { ValidatorsService } from 'services/validators.service';
 import { PromoBarBackendApiService } from 'services/promo-bar-backend-api.service';
 import { ExplorationDataService } from 'pages/exploration-editor-page/services/exploration-data.service';
 import { ExplorationDataBackendApiService } from 'pages/exploration-editor-page/services/exploration-data-backend-api.service';
+import { UserEmailPreferencesService } from 'pages/exploration-editor-page/services/user-email-preferences.service';
```
Please sort these files according to the path.

ashutoshc8101 replied:
Done

darksun27 reviewed (approved):
LGTM! Left a comment which is not a blocker.

ashutoshc8101 committed:
sorted

ashutoshc8101 committed:
Removed the empty line

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into UserEmailPreferencesServiceMigration

ashutoshc8101 commented:
I have updated my fork.

ashutoshc8101 committed:
requested changes by srijanreddy

ashutoshc8101 committed:
Requested changs

ashutoshc8101 committed:
changes

ashutoshc8101 committed:
sdf

ashutoshc8101 committed:
refactor

ashutoshc8101 committed:
fremove f

ashutoshc8101 committed:
100% coverge for settings-tab.component.ts

ashutoshc8101 committed:
f removed

ashutoshc8101 commented:
@srijanreddy98 
![Screenshot from 2021-02-26 15-19-17](https://user-images.githubusercontent.com/24855641/109284565-01e8a600-7846-11eb-9aca-8a141605eb5d.png)
The return type of `saveChangeToBackend` function is `Promise<void>` when http request is unsuccessfull
and `Promise<EmailPreferencesData>` when http request is successfull.
So, I found `Promise<void | EmailPreferencesData` to be the most appropriate return type for this function.
PTAL.

srijanreddy98 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences.service.ts:
```diff
@@ -16,77 +16,98 @@
+  suggestionNotificationsMuted: boolean;
+  constructor(
+    private alertsService: AlertsService,
+    private backendApiService: UserEmailPreferencesBackendApiService
```
rename this to `userEmailPreferencesBackendApiService`

ashutoshc8101 replied:
Done.

srijanreddy98 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences-backend-api.service.ts:
```diff
@@ -0,0 +1,53 @@
+
+  saveChangeToBackend(
+      requestParams: RequestParams
+  ): Promise<void | EmailPreferencesData> {
```
Change this to `Promise<EmailPreferencesData>`

ashutoshc8101 replied:
Done

srijanreddy98 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences.service.ts:
```diff
@@ -16,77 +16,98 @@
+   * @param {RequestParam} requestParams - Info about message_type and mute.
+   */
+  saveChangeToBackend(requestParams: RequestParams) :
+  Promise<void | EmailPreferencesData> {
```
Change this to `Promise<void>`

ashutoshc8101 replied:
Done

srijanreddy98 reviewed (changes requested):
Thanks @ashutoshc8101. I have a couple of nits otherwise LGTM!

ashutoshc8101 committed:
requested changes

ashutoshc8101 committed:
added empty lines

ashutoshc8101 reviewed (commented):
LGTM

ashutoshc8101 commented:
@srijanreddy98 
Thankyou for your review and suggestion.
I have implemented the suggested changes.
PTAL.

kevintab95 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences-backend-api.service.spec.ts:
```diff
@@ -0,0 +1,75 @@
+      req.flush(sampleResponse);
+      flushMicrotasks();
+      result.then((data : EmailPreferencesData) => {
+        expect(data).toBeDefined();
```
Can we explicitly check that data equals "sampleResponse"?

ashutoshc8101 replied:
Done.

kevintab95 commented on core/templates/pages/exploration-editor-page/services/user-email-preferences-backend-api.service.spec.ts:
```diff
@@ -0,0 +1,75 @@
+      flushMicrotasks();
+      result.then((data : EmailPreferencesData) => {
+        expect(data).toBeDefined();
+      }).then(done);
```
Is this required?

ashutoshc8101 replied:
No, it is not required.
Removing it.

kevintab95 reviewed (approved):
LGTM with 2 minor comments. Thanks @ashutoshc8101!

ashutoshc8101 committed:
changes

ashutoshc8101 commented:
@kevintab95 
Thankyou for your review and suggestions.
I have made the suggested changes.

sajalasati commented:
> @kevintab95
> Thankyou for your review and suggestions.
> I have made the suggested changes.

Hi @ashutoshc8101! Please always leave a comment saying `@reviewer PTAL` so that oppiabot automatically assigns the reviewer, otherwise you can yourself do the assignment if you want the reviewer to take a look.

Assigning @kevintab95 to take a look.

@srijanreddy98 PTAL!

ashutoshc8101 commented:
@sajalasati Thank you for your suggestion.
I made a comment just to inform him that the changes have been made.
As he had already approved the PR, there is no point in assigning him again.

srijanreddy98 reviewed (approved):
@ashutoshc8101 awesome work! LGTM!

ashutoshc8101 committed:
Merge remote-tracking branch 'upstream/develop' into UserEmailPreferencesServiceMigration

ashutoshc8101 commented:
@srijanreddy98 Thanks for the compliment.
